src/main/java/org/kiwiproject/dropwizard/util/startup/StartupLocker.java

```diff
@@ -0,0 +1,136 @@
+import lombok.Builder;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.recipes.locks.InterProcessLock;
+import org.apache.curator.framework.recipes.locks.InterProcessMutex;
+import org.kiwiproject.curator.CuratorFrameworkHelper;
+import org.kiwiproject.curator.CuratorLockHelper;
+import org.kiwiproject.curator.config.CuratorConfig;
+import org.kiwiproject.curator.exception.LockAcquisitionException;
+import org.kiwiproject.curator.zookeeper.ZooKeeperAvailabilityChecker;
+import org.kiwiproject.dropwizard.util.startup.PortAssigner.PortAssignment;
+import org.kiwiproject.dropwizard.util.startup.listener.StartupJettyLifeCycleListener;
+import org.kiwiproject.dropwizard.util.startup.listener.StartupWithLockJettyLifeCycleListener;
+
+import java.util.Optional;
+
+/**
+ * Utilities to acquire and release a lock from ZooKeeper during startup of the service
+ */
+@Slf4j
+public class StartupLocker {
+
+    @Getter(AccessLevel.PACKAGE)
+    @VisibleForTesting
+    private final ZooKeeperAvailabilityChecker zkAvailabilityChecker;
+
+    @Getter(AccessLevel.PACKAGE)
+    @VisibleForTesting
+    private final CuratorFrameworkHelper curatorFrameworkHelper;
+
+    @Getter(AccessLevel.PACKAGE)
+    @VisibleForTesting
+    private final CuratorLockHelper curatorLockHelper;
+
+    private final SystemExecutioner executioner;
+
+    @Builder
+    private StartupLocker(SystemExecutioner executioner,
+                          ZooKeeperAvailabilityChecker zkAvailabilityChecker,
+                          CuratorFrameworkHelper curatorFrameworkHelper,
+                          CuratorLockHelper curatorLockHelper) {
+        this.executioner = requireNotNull(executioner, "SystemExecutioner is required");
+        this.zkAvailabilityChecker = Optional.ofNullable(zkAvailabilityChecker).orElse(new ZooKeeperAvailabilityChecker());
+        this.curatorFrameworkHelper = Optional.ofNullable(curatorFrameworkHelper).orElse(new CuratorFrameworkHelper());
+        this.curatorLockHelper = Optional.ofNullable(curatorLockHelper).orElse(new CuratorLockHelper());
+    }
+
+    @AllArgsConstructor
+    public static class StartupLockInfo {
+        final CuratorFramework client;
+        final InterProcessLock lock;
+        final String lockPath;
+    }
+
+    /**
+     * Attempts to acquire a lock from ZooKeeper during startup.
+     *
+     * @param lockPath      the path in ZooKeeper to store the lock
+     * @param lockTimeout   the amount of time to wait for the lock to be acquired
+     * @param curatorConfig the Curator configuration
+     * @param environment   the Dropwizard environment
+     * @return information about the acquired lock or null if lock could not be acquired
+     */
+    public StartupLockInfo acquireStartupLock(String lockPath, Duration lockTimeout, PortAssignment assignment, CuratorConfig curatorConfig, Environment environment) {
```

Comment:
Maybe we take this opportunity to make this return `Optional`<StartupLockInfo>`?

We should then add some additional properties to `StartupLockInfo`:
* one to indicate if a lock was attempted (which would be false when using static port assignment)
* one to indicate if a lock was successfully obtained or not
* maybe an exception property in case there was a failure, e.g. a `LockAcquisitionException` is thrown in `tryAcquireStartupLock`
* another property to indicate something went wrong when an empty Optional is returned, but not due to an exception we caught. E.g. maybe there are no ZKs available. We would essentially want to put the WARN logging messages into this property, so that a client could (possibly) make a decision on how to proceed (if it wants to).

Does this make sense what I am proposing?

It would require some code changes to ensure exceptions don't escape, and maybe we'd also want to add a `hasException()` method in `StartupLockInfo` or something like that.

Reply:
I like this idea, however, with the additional properties you are suggesting, then Optional actually would never be empty because then you wouldn't have the properties.  I think we should expand out the `StartupLockInfo` object like you said but always return one.

Reply:
I've reworked this to have more information and never return null so I don't need the Optional